geps/gep-1897.md

```diff
@@ -0,0 +1,113 @@
+# GEP-1897: TLS from Gateway to Backend for ingress
+
+* Issue: [#1897](https://github.com/kubernetes-sigs/gateway-api/issues/1897)
+* Status: Provisional
+
+## TLDR
+
+This document specifically addresses the topic of conveying HTTPS from the Gateway
+dataplane to the backend (backend TLS termination), and intends to satisfy the single
+use case “As a client implementation of Gateway API, I need to know how to connect to
+a backend pod that has its own certificate”. TLS configuration can be a nebulous topic,
+so in order to drive resolution this GEP focuses only on this single piece of functionality.
+
+## Goals
+
+1. The solution should satisfy the following use case: the backend pod has its own
+certificate and the gateway implementation client needs to know how to connect to the
+backend pod. (Use case #4 in [Gateway API TLS Use Cases](#references))
+2. In terms of the Gateway API personas, the application developer should control the
+gateway to backend TLS settings, not the cluster operator, as requiring a cluster operator
+to manage certificate renewals and revocations would be extremely cumbersome.
+3. The solution should consider client certificate settings used in the TLS handshake **from
```

Comment:
I don't think TLS versions or cipher suites are appropriate - should be out of scope as well. 

They are typically restricted for compliance reasons  - an app owner should not be able to override. Also platform owner may enforce this by having infastructure restrict this ( compile only the compliant subset, etc). 

Another reason is that they require expertise that app owners typically don't have - very few people understand the compliance space.

Reply:
They are in scope for this discussion, as TLS version and cipher suites can be application dependent.  Whether the app owner can override the platform owner is an implementation detail too.

Reply:
My understanding:
* Apps will set allowed version/suites - this is not something we decide, that is how things work.
* Per costin, platform owners _should_ have a way to enforce gateway-api driven defaults.
* TLS version/cipher is a handshake

Where we may want a per-app/per-service gateway-api config override is if the platform admin says (for example) "TLS 1.3 is required", but an app says "I only speak TLS 1.2". As is, that would fail. However, we could allow a per-service opt-out of the platform rules and say "for only svc-legacy, allow 1.2".

This seems a bit strange to me, though. If we want TLS 1.3 everywhere, we probably _really_ want it everywhere. Its not clear why we would be ok with less secure methods for one service but not another.

If we just want to default to more secure options, that is very different - as mentioned, TLS is a handshake negotiation on these, so we can say at admin level "Support 1.3,1.2 and suites [best,2ndbest,3rdbest]" or equivilent.

Reply:
I agree that an override or default should be possible.  However, I will leave this discussion for another GEP or the implementation details.  Every cluster has its use cases, and we may need to address a cluster-level override or default with a different set of use cases.  Sound fair?

Reply:
I don't think this is an "overrides and defaults" scenario. We need that when we are talking about something like Retry or Timeout. TLS is different; it is an existing protocol which already has a negotiation aspect. We cannot and should not invent another negotiation on top of that.

Reply:
+1

Also UX is important - FIPS compliance and what cipher suites are safe or required is a completely different persona than Jane the app developer.  It is something that may be implemented by compile options and deep into platform - and if a cipher or TLS version becomes unsafe it is not up to the developers to make changes, but platform owner.

Reply:
@howardjohn and @costinm, everything you've said is correct, but you are missing the function that end users end up using an ingress proxy for - fixing broken apps. It's _very_ common that the TLS version policy will be "Everything must be TLS 1.3, except for this list of exceptions". Having a way to describe that list of exceptions is imperative.

Yes, this can absolutely be a footgun, but it is _definitely_ required in real-world use cases. We had to add similar functionality to Contour for exactly this reason.

Does this create effectively another negotiation over the top of the TLS one? Yes. Is it still unfortunately necessary? Also yes.

This is just one of those things that people who are stuck working in a service-mesh free world are stuck with.

Reply:
Considering this resolved as there hasn't been an update in some time. If you feel there's still action that needs to be taken here please unresolve and let us know in a comment.